### PR TITLE
feat: Auth0 Components

### PR DESCRIPTION
**Pull request by milldr**
## what
- Added components for Auth0 terraformation - `auth0/tenant` and `auth0/app`

## why
- Auth0 Application component. Auth0 is a third-party service that provides authentication and authorization as a service. It is typically used to to authenticate users.
- `auth0/tenant` configures the Terraform provider and the Auth0 tenant itself
- `auth0/app` deploys an Auth0 application

## references
- client engagement

### CODE REVIEW

File: modules/auth0/app/outputs.tf
Comment (milldr) [inline modules/auth0/app/outputs.tf:1]: add some output

File: modules/auth0/app/README.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
+# Component: `auth0/app`
+
+Auth0 Application component. Auth0 is a third-party service that provides authentication and authorization as a service.
```
Comment (korenyoni) [inline modules/auth0/app/README.md:3]: nit: add link to Auth0's landing page or documentation landing page. Not a must but just nice to have in the README IMO.

File: modules/auth0/app/README.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+## Outputs
+
+| Name | Description |
+|------|-------------|
+| <a name="output_auth0_client_id"></a> [auth0\_client\_id](#output\_auth0\_client\_id) | The ID of the Auth0 client |
+<!-- END OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
+<!-- prettier-ignore-end -->
+
+## References
```
Comment (korenyoni) [inline modules/auth0/app/README.md:130]: Maybe add the Auth0 links here.

File: modules/auth0/app/outputs.tf [hunk 3]
```diff
@@ -0,0 +1,4 @@
+output "auth0_client_id" {
+  value       = auth0_client.this[0].client_id
+  description = "The ID of the Auth0 client"
```
Comment (korenyoni) [inline modules/auth0/app/outputs.tf:3]: nit

```suggestion
  description = "The Auth0 Application Client ID"
```

File: modules/auth0/app/provider-auth0-client.tf [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  default     = ""
+}
+
+variable "auth0_tenant_stage_name" {
+  description = "The name of the stage where the Auth0 tenant component is deployed. Defaults to the stage of the current stack."
+  type        = string
+  default     = ""
+}
+
+variable "auth0_tenant_tenant_name" {
```
Comment (korenyoni) [inline modules/auth0/app/provider-auth0-client.tf:22]: Could we avoid the double `tenant`?

File: modules/auth0/app/provider-auth0-client.tf [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+#
+# Initialize the Auth0 provider with the Auth0 domain, client ID, and client secret from that deployment
+#
+provider "auth0" {
+  domain        = data.aws_ssm_parameter.auth0_domain.value
+  client_id     = data.aws_ssm_parameter.auth0_client_id.value
+  client_secret = data.aws_ssm_parameter.auth0_client_secret.value
+  debug         = true
```
Comment (korenyoni) [inline modules/auth0/app/provider-auth0-client.tf:90]: Should this always be set to `true`?

File: modules/auth0/tenant/README.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
+# Component: `auth0/tenant`
```
Comment (korenyoni) [inline modules/auth0/tenant/README.md:1]: See README suggestions from `app` component

File: modules/auth0/tenant/main.tf [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+#
+# However, we can import it using a random string. On first run, we import the existing tenant
+# using a random string. It does not matter what this value is. Terraform will use the same
+# tenant as the Auth0 application for the Terraform Auth0 Provider.
+#
+# https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/tenant#import
+import {
+  to = auth0_tenant.this[0]
+  id = "f6615002-81ff-49f7-afd3-8814d07af4fa"
+}
```
Comment (korenyoni) [inline modules/auth0/tenant/main.tf:23]: I think this is a leftover from development work

File: modules/auth0/tenant/main.tf [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  default_redirection_uri = var.default_redirection_uri
+
+  flags {
+    disable_clickjack_protection_headers   = var.disable_clickjack_protection_headers
+    enable_public_signup_user_exists_error = var.enable_public_signup_user_exists_error
+    use_scope_descriptions_for_consent     = var.use_scope_descriptions_for_consent
+    no_disclose_enterprise_connections     = var.no_disclose_enterprise_connections
+    disable_management_api_sms_obfuscation = var.disable_management_api_sms_obfuscation
+    disable_fields_map_fix                 = var.disable_fields_map_fix
+  }
```
Comment (korenyoni) [inline modules/auth0/tenant/main.tf:47]: Maybe these would be better as a single map variable, but not a must.

File: modules/auth0/tenant/main.tf [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  domain = local.domain_name
+  type   = "auth0_managed_certs"
+}
+
+resource "aws_route53_record" "this" {
+  count = local.enabled ? 1 : 0
+
+  zone_id = module.dns_gbl_delegated.outputs.default_dns_zone_id
+  name    = local.domain_name
+  type    = upper(auth0_custom_domain.this[0].verification[0].methods[0].name)
```
Comment (korenyoni) [inline modules/auth0/tenant/main.tf:70]: Is this an issue when `local.enabled` is false? i.e. maybe we need `join()` combined with a glob star?

File: modules/auth0/tenant/main.tf [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  ttl     = "300"
+  records = [
+    auth0_custom_domain.this[0].verification[0].methods[0].record,
```
Comment (korenyoni) [inline modules/auth0/tenant/main.tf:73]: Same here

File: modules/auth0/tenant/variables.tf [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+variable "friendly_name" {
+  type        = string
+  description = "The friendly name of the Auth0 tenant. If not provided, the module context ID will be used."
+  default     = ""
+}
+
+variable "picture_url" {
+  type        = string
+  description = "The URL of the picture to be displayed in the Auth0 Universal Login page."
+  default     = "" # TODO provide a default here
```
Comment (korenyoni) [inline modules/auth0/tenant/variables.tf:15]: See TODO

**Review by korenyoni — COMMENTED** [review]
Overall looks really good. Just caught a couple of small issues.

File: modules/auth0/tenant/variables.tf (hunk 11 above)
Comment (korenyoni) [reply modules/auth0/tenant/variables.tf:15]: I'd suggest either linking to the Cloud Posse logo, probably

File: modules/auth0/app/provider-auth0-client.tf (hunk 4 above)
Comment (milldr) [reply modules/auth0/app/provider-auth0-client.tf:22]: plus we use the same pattern frequently for remote state component context

```
<component-name>_tenant_name
<component-name>_stage_name
<component-name>_environment_name
```

But the component name here is `auth0_tenant`

File: modules/auth0/tenant/main.tf (hunk 7 above)
Comment (milldr) [reply modules/auth0/tenant/main.tf:23]: It's a workaround for a nonexistent identifier in Auth0

See [this comment above
](https://github.com/cloudposse/terraform-aws-components/blob/7aac04bb96d00d64c59e15410e3e52c2d975c24e/modules/auth0/tenant/main.tf#L10-L19)

And the README: https://github.com/cloudposse/terraform-aws-components/blob/7aac04bb96d00d64c59e15410e3e52c2d975c24e/modules/auth0/tenant/README.md#auth0-tenant-creation

File: modules/auth0/app/README.md (hunk 2 above)
Comment (milldr) [reply modules/auth0/app/README.md:130]: added

File: modules/auth0/app/README.md (hunk 1 above)
Comment (milldr) [reply modules/auth0/app/README.md:3]: added hyperlinks

File: modules/auth0/app/provider-auth0-client.tf (hunk 5 above)
Comment (milldr) [reply modules/auth0/app/provider-auth0-client.tf:90]: Adding a var

File: modules/auth0/tenant/README.md (hunk 6 above)
Comment (milldr) [reply modules/auth0/tenant/README.md:1]: and reference links

File: modules/auth0/tenant/main.tf (hunk 8 above)
Comment (milldr) [reply modules/auth0/tenant/main.tf:47]: Personally I prefer them each as a variable. That way you cannot unintentionally overwrite the map with merge variables (even though maps are deep merged).

Plus the `flag` block should only be defined once, so each input is also only defined once

File: modules/auth0/tenant/main.tf (hunk 9 above)
Comment (milldr) [reply modules/auth0/tenant/main.tf:70]: good catch; that's likely an issue. adding `join()`

File: modules/auth0/tenant/main.tf (hunk 10 above)
Comment (milldr) [reply modules/auth0/tenant/main.tf:73]: resolved as well

File: modules/auth0/tenant/main.tf (hunk 9 above)
Comment (milldr) [reply modules/auth0/tenant/main.tf]: handled `enabled: false` for `auth0/tenant`

File: modules/auth0/tenant/variables.tf (hunk 11 above)
Comment (milldr) [reply modules/auth0/tenant/variables.tf:15]: added a logo

File: modules/auth0/app/provider-auth0-client.tf (hunk 4 above)
Comment (korenyoni) [reply modules/auth0/app/provider-auth0-client.tf:22]: Understandable, it just confused me when I saw it but makes total sense and I agree.

File: modules/auth0/tenant/main.tf (hunk 7 above)
Comment (korenyoni) [reply modules/auth0/tenant/main.tf:23]: Ah got it. Missed that.

File: modules/auth0/app/variables.tf [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+variable "grant_types" {
+  type        = list(string)
+  description = "Allowed Grant Types"
+  default     = []
+}
+
+variable "logo_uri" {
+  type        = string
+  description = "Logo URI"
+  default     = "" # TODO Add default logo
```
Comment (korenyoni) [inline modules/auth0/app/variables.tf:33]: @milldr I'm still seeing the TODO here — maybe you didn't push it?

**Review by korenyoni — APPROVED** [review]
LGTM

**Pull request merged**